kafkarator/writer.go

```diff
@@ -3,70 +3,111 @@ package kafkarator
 import (
 	"context"
 	"fmt"
+	"maps"
 
+	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
 	"github.com/hafslundkraft/golib/telemetry"
-	"github.com/segmentio/kafka-go"
 	"go.opentelemetry.io/otel"
 	"go.opentelemetry.io/otel/metric"
 	"go.opentelemetry.io/otel/propagation"
 )
 
-func newWriter(w *kafka.Writer, pmc metric.Int64Counter, tel *telemetry.Provider) *Writer {
-	return &Writer{
-		writer:                  w,
+func newWriter(
+	p *kafka.Producer,
+	topic string,
+	pmc metric.Int64Counter,
+	tel *telemetry.Provider,
+) *Writer {
+	w := &Writer{
+		producer:                p,
+		topic:                   topic,
 		tel:                     tel,
 		producedMessagesCounter: pmc,
 		closed:                  false,
 	}
+
+	go w.handleDeliveryReports()
+
+	return w
 }
 
 // Writer provides an interface for writing messages to the Kafka topic, as well
 // as closing it when the client is done writing.
 type Writer struct {
 	producedMessagesCounter metric.Int64Counter
-	writer                  *kafka.Writer
+	producer                *kafka.Producer
+	topic                   string
 	tel                     *telemetry.Provider
 	closed                  bool
 }
 
 // Close closes the underlying infrastructure, and renders this interface unusable for writing messages.
-func (wc *Writer) Close(ctx context.Context) error {
-	if wc.closed {
+func (w *Writer) Close(ctx context.Context) error {
+	if w.closed {
 		return nil // It's ok to close multiple times.
 	}
 
-	if err := wc.writer.Close(); err != nil {
-		wc.tel.Logger().ErrorContext(ctx, fmt.Sprintf("failed to close writer %v", err))
-		return fmt.Errorf("failed to close writer: %w", err)
-	}
-	wc.closed = true
+	w.producer.Flush(5000)
+	w.producer.Close()
+
+	w.closed = true
+
 	return nil
 }
 
 // Write writes the given message with headers to the topic. An important side effect is
 // that if there is an OpenTelemetry tracing span associated with the context, it is extracted
 // and included in the header that is sent to Kafka.
-func (wc *Writer) Write(ctx context.Context, msg []byte, headers map[string][]byte) error {
-	if wc.closed {
+func (w *Writer) Write(ctx context.Context, key, value []byte, headers map[string][]byte) error {
```

Comment:
Kan det være en ide å pakke disse verdiene i en kafkarator.Message, for å unngå at det blir en parameter-suppe her? Det er jo flere elementer som kan tas inn som parametere her på sikt, og dersom det ikke er pakket i en message så vil man brekke grensesnittet hvis man legger til noe.